ocaml/tests/test_observer.ml

```diff
@@ -55,35 +139,209 @@ let start_test_trace () =
+  TracerProvider.assert_num_observers ~__context 1 ;
+
+  test_destroy ~__context ~self () ;
+  Xapi_DB.assert_num_observers ~__context 0 ;
+  TracerProvider.assert_num_observers ~__context 0
+
+let test_hosts ~__context ~self =
+  let valid_host = !Xapi_globs.localhost_ref in
+  ( try Xapi_observer.set_hosts ~__context ~self ~value:[valid_host]
+    with _ -> Alcotest.failf "Failed to set valid host"
+  ) ;
+
+  let invalid_host = Ref.null in
+  Alcotest.check_raises "Xapi_observer.set_hosts should fail on invalid host"
+    Api_errors.(
+      Server_error (invalid_value, ["host"; Ref.string_of invalid_host])
+    )
+    (fun () ->
+      Xapi_observer.set_hosts ~__context ~self ~value:[invalid_host] |> ignore
+    )
+
+let test_components ~__context ~self =
+  let valid_components = ["xapi"] in
+  List.iter
+    (fun x ->
+      try Xapi_observer.set_components ~__context ~self ~value:[x]
+      with _ -> Alcotest.failf "Failed to set valid component"
+    )
+    valid_components ;
+
+  let invalid_component = "invalid-component" in
+  Alcotest.check_raises
+    "Xapi_observer.set_components should fail on invalid component"
+    Api_errors.(Server_error (invalid_value, ["component"; invalid_component]))
+    (fun () ->
+      Xapi_observer.set_components ~__context ~self ~value:[invalid_component]
+      |> ignore
+    )
+
+let test_endpoints ~__context ~self =
+  let valid_endpoints = ["bugtool"; "http://example.com:9411/api/v2/spans"] in
+  List.iter
+    (fun x ->
+      try Xapi_observer.set_endpoints ~__context ~self ~value:[x]
+      with _ -> Alcotest.failf "Failed to set valid endpoint"
+    )
+    valid_endpoints
+
+let test_observer_valid_params () =
+  let __context = Test_common.make_test_database () in
+  let self = test_create ~__context ~name_label:"test-observer" () in
+
+  test_hosts ~__context ~self ;
+  test_components ~__context ~self ;
+  test_endpoints ~__context ~self ;
+
+  test_destroy ~__context ~self ()
+
+let test_observer_endpoint () =
+  let __context = Test_common.make_test_database () in
+  let self = test_create ~__context () in
+  let url = "http://example.com:9411/api/v2/spans" in
+  let endpoints = [url] in
+  Xapi_observer.set_endpoints ~__context ~self ~value:endpoints ;
+  Xapi_DB.check_endpoints ~__context ~self ~endpoints ;
+  TracerProvider.check_endpoints ~name:"test-observer" ~endpoints ;
+
+  let endpoints = ["bugtool"; url] in
+  Xapi_observer.set_endpoints ~__context ~self ~value:endpoints ;
+  Xapi_DB.check_endpoints ~__context ~self ~endpoints ;
+  TracerProvider.check_endpoints ~name:"test-observer" ~endpoints ;
+  test_destroy ~__context ~self ()
+
+let is_export_dir_empty () =
+  try
+    let log_dir =
+      Sys.readdir test_trace_log_dir
+      |> Array.map (Filename.concat test_trace_log_dir)
+    in
+    Array.length log_dir = 0
+  with e ->
+    debug "%s" (Printexc.to_string e) ;
+    true
+
+let clear_export_dir () =
+  try
+    let rec rmdir_rec path =
```

Comment:
Generating a temporary folder for each test is better than reusing a single one for all of them, it may allow to run them in parallel and avoid these cleanup functions